### ISSUE TITLE
Trouble with HTTPS/TLS termination with use-proxy-protocol

### ISSUE THREAD

**Issue by anapsix** [heptio/contour#793]
**TL;DR**

Unable to reach Ingress over HTTPS.
Chrome Browser reports `ERR_CONNECTION_CLOSED `
Curl returns `SSL_ERROR_SYSCALL` or `Server aborted the SSL handshake` depending on the version.

Debug logs from Envoy shows  `closing connection: no matching filter chain found`

**What steps did you take and what happened:**

Following tutorial, deployed Contour as DaemonSet on VM-based cluster.
Enabled proxy-protocol, changed http/https ports to 9080 and 9443 respectively.
```
- image: gcr.io/heptio-images/contour:master
  imagePullPolicy: Always
  name: contour
  command: ["contour"]
  args:
  - serve
  - --incluster
  - --envoy-http-port=9080
  - --envoy-https-port=9443
  - --use-proxy-protocol
```

Did not want to use `hostNetwork: true` to avoid exposing other Contour / Envoy ports
```
$ kubectl -n heptio-contour exec contour-znq2k -c envoy -- sh -c 'netstat -lnpt'
netstat: /proc/net/tcp6: No such file or directory
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name
tcp        0      0 0.0.0.0:8002            0.0.0.0:*               LISTEN      1/envoy
tcp        0      0 0.0.0.0:9443            0.0.0.0:*               LISTEN      1/envoy
tcp        0      0 127.0.0.1:9001          0.0.0.0:*               LISTEN      1/envoy
tcp        0      0 127.0.0.1:6060          0.0.0.0:*               LISTEN      -
tcp        0      0 0.0.0.0:9080            0.0.0.0:*               LISTEN      1/envoy
tcp        0      0 0.0.0.0:8000            0.0.0.0:*               LISTEN      -
tcp        0      0 127.0.0.1:8001          0.0.0.0:*               LISTEN      -
```

Configured envoy container with hostPort 9080 / 9443 for http and https respectively.
```
- image: docker.io/envoyproxy/envoy-alpine:v1.7.0
  name: envoy
  ports:
    - name: http
      hostPort: 9080
      containerPort: 9080
    - name: https
      hostPort: 9443
      containerPort: 9443
  command: ["envoy"]
    args:
      - --config-path /config/contour.yaml
      - --service-cluster cluster0
      - --service-node node0
      - --log-level info
      - --v2-config-only
```
Deployed cert-manager, with [Self Signing Issuer](https://cert-manager.readthedocs.io/en/latest/reference/issuers/selfsigned/index.html), and [CA Issuer](https://cert-manager.readthedocs.io/en/latest/tutorials/ca/creating-ca-issuer.html). Both are working as expected, generating certificates when Ingress contains `certmanager.k8s.io/cluster-issuer` annotation.

Deployed Kuard example with simple Ingress
```
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: kuard
  annotations:
    certmanager.k8s.io/cluster-issuer: self-signed-CA
    ingress.kubernetes.io/force-ssl-redirect: "false"
  labels:
    app: kuard
spec:
  rules:
    - host: kuard.random.io
      http:
        paths:
        - backend:
            serviceName: kuard
            servicePort: 80
  tls:
    - hosts:
      - kuard.random.io
      secretName: kuard-cert
```
Certificate and TLS secret got created, 

**What did you expect to happen:**

Expected to be able to reach HTTPS ingress, just as easily as HTTP.

**Anything else you would like to add:**

Checking with CURL and `--haproxy-protocol`.  
HTTP:
```
## via SSH to K8s node, with port forwarding
## and /etc/hosts: 127.0.0.1 kuard.random.io
$ /usr/local/Cellar/curl/7.62.0/bin/curl --haproxy-protocol -o/dev/null -4 -sv -H 'Host: kuard.random.io' http://kuard.random.io:9080
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to kuard.random.io (127.0.0.1) port 9080 (#0)
> PROXY TCP4 127.0.0.1 127.0.0.1 51624 9080
> GET / HTTP/1.1
> Host: kuard.random.io
> User-Agent: Mozilla/5.0 (compatible; MSIE 9.0; Windows NT 6.1; Trident/5.0)
> Accept: */*
> Referer:
>
< HTTP/1.1 200 OK
< content-length: 1767
< content-type: text/html
< date: Wed, 31 Oct 2018 19:13:51 GMT
< x-envoy-upstream-service-time: 1
< server: envoy
<
{ [1767 bytes data]
* Connection #0 to host kuard.random.io left intact

## from envoy container
$ kubectl -n heptio-contour exec contour-znq2k -c envoy -- sh -c 'apk add --no-cache -q curl && curl --haproxy-protocol -o/dev/null -H "Host: kuard.random.io" -vsk http://localhost:9080'
* Rebuilt URL to: http://localhost:9080/
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to localhost (127.0.0.1) port 9080 (#0)
> PROXY TCP4 127.0.0.1 127.0.0.1 32838 9080
> GET / HTTP/1.1
> Host: kuard.random.io
> User-Agent: curl/7.61.1
> Accept: */*
>
< HTTP/1.1 200 OK
< content-length: 1702
< content-type: text/html
< date: Wed, 31 Oct 2018 17:27:42 GMT
< x-envoy-upstream-service-time: 1
< server: envoy
<
{ [1702 bytes data]
* Connection #0 to host localhost left intact
```

HTTPS:
```
## using openssl
$ /usr/local/Cellar/openssl/1.0.2p/bin/openssl s_client -connect 127.0.0.1:9443 -servername kuard.random.io
CONNECTED(00000003)
4692285036:error:140790E5:SSL routines:ssl23_write:ssl handshake failure:s23_lib.c:177:
---
no peer certificate available
---
No client certificate CA names sent
---
SSL handshake has read 0 bytes and written 337 bytes
---
New, (NONE), Cipher is (NONE)
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
SSL-Session:
    Protocol  : TLSv1.2
    Cipher    : 0000
    Session-ID:
    Session-ID-ctx:
    Master-Key:
    Key-Arg   : None
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    Start Time: 1541015117
    Timeout   : 300 (sec)
    Verify return code: 0 (ok)
---

## via SSH to K8s node, with port forwarding
## and /etc/hosts: 127.0.0.1 kuard.random.io
/usr/local/Cellar/curl/7.62.0/bin/curl --haproxy-protocol -o/dev/null -4 --cacert ../cluster_resources/helm/cert-manager/ssl/ca.crt -sv https://kuard.random.io:9443
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to kuard.random.io (127.0.0.1) port 9443 (#0)
> PROXY TCP4 127.0.0.1 127.0.0.1 51611 9443
* ALPN, offering http/1.1
* Server aborted the SSL handshake
* Closing connection 0

## from envoy container
$ kubectl -n heptio-contour exec contour-znq2k -c envoy -- sh -c 'apk add --no-cache -q curl && curl --haproxy-protocol -o/dev/null -H "Host: kuard.random.io" -vsk https://localhost:9443'
* Rebuilt URL to: https://localhost:9443/
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to localhost (127.0.0.1) port 9443 (#0)
> PROXY TCP4 127.0.0.1 127.0.0.1 33762 9443
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/certs/ca-certificates.crt
  CApath: none
* TLSv1.2 (OUT), TLS handshake, Client hello (1):
} [224 bytes data]
* LibreSSL SSL_connect: SSL_ERROR_SYSCALL in connection to localhost:9443
command terminated with exit code 35
```

Debug logs from Envoy contain the following
```
[2018-10-31 19:05:24.541][14][debug][filter] source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc:28] proxy_protocol: New connection accepted
[2018-10-31 19:05:26.179][15][debug][filter] source/extensions/filters/listener/tls_inspector/tls_inspector.cc:73] tls inspector: new connection accepted
[2018-10-31 19:05:26.180][15][debug][filter] source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc:28] proxy_protocol: New connection accepted
[2018-10-31 19:05:26.180][15][debug][main] source/server/connection_handler_impl.cc:192] closing connection: no matching filter chain found
[2018-10-31 19:05:27.830][17][debug][filter] source/extensions/filters/listener/tls_inspector/tls_inspector.cc:73] tls inspector: new connection accepted
```

**Environment:**

- Contour version: gcr.io/heptio-images/contour:master
- Kubernetes version: (use `kubectl version`): v1.11.2
- Kubernetes installer & version: kubeadm v1.11.2
- Cloud provider or hardware configuration: VMs on bare-metal
- OS (e.g. from `/etc/os-release`): CentOS 7

**Comment by stevesloka** [comment]
Hey @anapsix, did you deploy the service as well? There are some expectations that Contour configures. I'm curious if you curl the service over `:443` to see if it works. 

Also, could you go to the Envoy admin page on port 9001 and verify the https listener got created?

**Comment by davecheney** [comment]
My guess is the SNI handshake the client sent is not matching the list that envoy expects, can you please  follow

https://github.com/heptio/contour/blob/master/docs/troubleshooting.md#access-the-envoy-admin-interface-remotely

and paste the contents of 

http://127.0.0.1:9001/config_dump

## NOTE: This dump contains SSL key and cert data, please anonymise it if this is sensitive.

**Comment by anapsix** [comment]
@stevesloka, yep service for `kuard` deployed. But it would not be listening on 443, since I'm looking for Contour Ingress to provide TLS termination.
```
apiVersion: v1
kind: Service
metadata:
  labels:
    app: kuard
  name: kuard
spec:
  ports:
  - port: 80
    protocol: TCP
    targetPort: 8080
  selector:
    app: kuard
  sessionAffinity: None
  type: ClusterIP
```
For Contour, I'm using DaemonSet with `hostPort`. 

----
@davecheney that's what I thought.. I was hoping `openssl s_client -connect 127.0.0.1:9443 -servername kuard.random.io` would help me catch that

Here is the config dump
```
{
  "configs": {
    "bootstrap": {
      "@type": "type.googleapis.com/envoy.admin.v2alpha.BootstrapConfigDump",
      "bootstrap": {
        "node": {
          "buildVersion": "3f59fb5c0f6554f8b3f2e73ab4c1437a63d42668/1.7.0/Clean/RELEASE"
        },
        "staticResources": {
          "listeners": [
            {
              "address": {
                "socketAddress": {
                  "address": "0.0.0.0",
                  "portValue": 8002
                }
              },
              "filterChains": [
                {
                  "filters": [
                    {
                      "name": "envoy.http_connection_manager",
                      "config": {
                        "route_config": {
                          "virtual_hosts": [
                            {
                              "routes": [
                                {
                                  "route": {
                                    "cluster": "service_stats"
                                  },
                                  "match": {
                                    "prefix": "/stats"
                                  }
                                }
                              ],
                              "domains": [
                                "*"
                              ],
                              "name": "backend"
                            }
                          ]
                        },
                        "codec_type": "AUTO",
                        "http_filters": [
                          {
                            "name": "envoy.health_check",
                            "config": {
                              "endpoint": "/healthz",
                              "pass_through_mode": false
                            }
                          },
                          {
                            "name": "envoy.router",
                            "config": null
                          }
                        ],
                        "stat_prefix": "stats"
                      }
                    }
                  ]
                }
              ]
            }
          ],
          "clusters": [
            {
              "name": "contour",
              "type": "STRICT_DNS",
              "connectTimeout": "5s",
              "hosts": [
                {
                  "socketAddress": {
                    "address": "127.0.0.1",
                    "portValue": 8001
                  }
                }
              ],
              "circuitBreakers": {
                "thresholds": [
                  {
                    "priority": "HIGH",
                    "maxConnections": 100000,
                    "maxPendingRequests": 100000,
                    "maxRequests": 60000000,
                    "maxRetries": 50
                  },
                  {
                    "maxConnections": 100000,
                    "maxPendingRequests": 100000,
                    "maxRequests": 60000000,
                    "maxRetries": 50
                  }
                ]
              },
              "http2ProtocolOptions": {}
            },
            {
              "name": "service_stats",
              "type": "LOGICAL_DNS",
              "connectTimeout": "0.250s",
              "hosts": [
                {
                  "socketAddress": {
                    "address": "127.0.0.1",
                    "portValue": 9001
                  }
                }
              ]
            }
          ]
        },
        "dynamicResources": {
          "ldsConfig": {
            "apiConfigSource": {
              "apiType": "GRPC",
              "grpcServices": [
                {
                  "envoyGrpc": {
                    "clusterName": "contour"
                  }
                }
              ]
            }
          },
          "cdsConfig": {
            "apiConfigSource": {
              "apiType": "GRPC",
              "grpcServices": [
                {
                  "envoyGrpc": {
                    "clusterName": "contour"
                  }
                }
              ]
            }
          }
        },
        "admin": {
          "accessLogPath": "/dev/null",
          "address": {
            "socketAddress": {
              "address": "127.0.0.1",
              "portValue": 9001
            }
          }
        }
      }
    },
    "routes": {
      "@type": "type.googleapis.com/envoy.admin.v2alpha.RoutesConfigDump",
      "staticRouteConfigs": [
        {
          "virtualHosts": [
            {
              "name": "backend",
              "domains": [
                "*"
              ],
              "routes": [
                {
                  "match": {
                    "prefix": "/stats"
                  },
                  "route": {
                    "cluster": "service_stats"
                  }
                }
              ]
            }
          ]
        }
      ],
      "dynamicRouteConfigs": [
        {
          "versionInfo": "0",
          "routeConfig": {
            "name": "ingress_https",
            "virtualHosts": [
              {
                "name": "kuard.random.io",
                "domains": [
                  "kuard.random.io",
                  "kuard.random.io:443"
                ],
                "routes": [
                  {
                    "match": {
                      "prefix": "/"
                    },
                    "route": {
                      "cluster": "default/kuard/80/da39a3ee5e",
                      "requestHeadersToAdd": [
                        {
                          "header": {
                            "key": "x-request-start",
                            "value": "t=%START_TIME(%s.%3f)%"
                          },
                          "append": true
                        }
                      ]
                    }
                  }
                ]
              }
            ]
          }
        },
        {
          "versionInfo": "0",
          "routeConfig": {
            "name": "ingress_http",
            "virtualHosts": [
              {
                "name": "kuard.random.io",
                "domains": [
                  "kuard.random.io",
                  "kuard.random.io:80"
                ],
                "routes": [
                  {
                    "match": {
                      "prefix": "/"
                    },
                    "route": {
                      "cluster": "default/kuard/80/da39a3ee5e",
                      "requestHeadersToAdd": [
                        {
                          "header": {
                            "key": "x-request-start",
                            "value": "t=%START_TIME(%s.%3f)%"
                          },
                          "append": true
                        }
                      ]
                    }
                  }
                ]
              }
            ]
          }
        }
      ]
    },
    "listeners": {
      "@type": "type.googleapis.com/envoy.admin.v2alpha.ListenersConfigDump",
      "versionInfo": "0",
      "staticListeners": [
        {
          "address": {
            "socketAddress": {
              "address": "0.0.0.0",
              "portValue": 8002
            }
          },
          "filterChains": [
            {
              "filters": [
                {
                  "name": "envoy.http_connection_manager",
                  "config": {
                    "route_config": {
                      "virtual_hosts": [
                        {
                          "routes": [
                            {
                              "route": {
                                "cluster": "service_stats"
                              },
                              "match": {
                                "prefix": "/stats"
                              }
                            }
                          ],
                          "domains": [
                            "*"
                          ],
                          "name": "backend"
                        }
                      ]
                    },
                    "codec_type": "AUTO",
                    "http_filters": [
                      {
                        "name": "envoy.health_check",
                        "config": {
                          "endpoint": "/healthz",
                          "pass_through_mode": false
                        }
                      },
                      {
                        "name": "envoy.router",
                        "config": null
                      }
                    ],
                    "stat_prefix": "stats"
                  }
                }
              ]
            }
          ]
        }
      ],
      "dynamicActiveListeners": [
        {
          "versionInfo": "0",
          "listener": {
            "name": "ingress_http",
            "address": {
              "socketAddress": {
                "address": "0.0.0.0",
                "portValue": 9080
              }
            },
            "filterChains": [
              {
                "filters": [
                  {
                    "name": "envoy.http_connection_manager",
                    "config": {
                      "http_protocol_options": {
                        "accept_http_10": true
                      },
                      "use_remote_address": true,
                      "stat_prefix": "ingress_http",
                      "rds": {
                        "route_config_name": "ingress_http",
                        "config_source": {
                          "api_config_source": {
                            "api_type": "GRPC",
                            "grpc_services": [
                              {
                                "envoy_grpc": {
                                  "cluster_name": "contour"
                                }
                              }
                            ]
                          }
                        }
                      },
                      "access_log": [
                        {
                          "name": "envoy.file_access_log",
                          "config": {
                            "path": "/dev/stdout"
                          }
                        }
                      ],
                      "http_filters": [
                        {
                          "name": "envoy.gzip"
                        },
                        {
                          "name": "envoy.grpc_web"
                        },
                        {
                          "name": "envoy.router"
                        }
                      ]
                    }
                  }
                ],
                "useProxyProto": true
              }
            ]
          }
        },
        {
          "versionInfo": "0",
          "listener": {
            "name": "ingress_https",
            "address": {
              "socketAddress": {
                "address": "0.0.0.0",
                "portValue": 9443
              }
            },
            "filterChains": [
              {
                "filterChainMatch": {
                  "serverNames": [
                    "kuard.random.io"
                  ]
                },
                "tlsContext": {
                  "commonTlsContext": {
                    "tlsParams": {
                      "tlsMinimumProtocolVersion": "TLSv1_1",
                      "tlsMaximumProtocolVersion": "TLSv1_3"
                    },
                    "tlsCertificates": [
                      {
                        "certificateChain": {
                          "inlineBytes": "REDACTED"
                        },
                        "privateKey": {
                          "inlineBytes": "REDACTED"
                        }
                      }
                    ],
                    "alpnProtocols": [
                      "h2",
                      "http/1.1"
                    ]
                  }
                },
                "filters": [
                  {
                    "name": "envoy.http_connection_manager",
                    "config": {
                      "http_protocol_options": {
                        "accept_http_10": true
                      },
                      "rds": {
                        "route_config_name": "ingress_https",
                        "config_source": {
                          "api_config_source": {
                            "api_type": "GRPC",
                            "grpc_services": [
                              {
                                "envoy_grpc": {
                                  "cluster_name": "contour"
                                }
                              }
                            ]
                          }
                        }
                      },
                      "use_remote_address": true,
                      "stat_prefix": "ingress_https",
                      "access_log": [
                        {
                          "config": {
                            "path": "/dev/stdout"
                          },
                          "name": "envoy.file_access_log"
                        }
                      ],
                      "http_filters": [
                        {
                          "name": "envoy.gzip"
                        },
                        {
                          "name": "envoy.grpc_web"
                        },
                        {
                          "name": "envoy.router"
                        }
                      ]
                    }
                  }
                ],
                "useProxyProto": true
              }
            ],
            "listenerFilters": [
              {
                "name": "envoy.listener.tls_inspector",
                "config": {}
              }
            ]
          }
        }
      ]
    },
    "clusters": {
      "@type": "type.googleapis.com/envoy.admin.v2alpha.ClustersConfigDump",
      "versionInfo": "0",
      "staticClusters": [
        {
          "name": "contour",
          "type": "STRICT_DNS",
          "connectTimeout": "5s",
          "hosts": [
            {
              "socketAddress": {
                "address": "127.0.0.1",
                "portValue": 8001
              }
            }
          ],
          "circuitBreakers": {
            "thresholds": [
              {
                "priority": "HIGH",
                "maxConnections": 100000,
                "maxPendingRequests": 100000,
                "maxRequests": 60000000,
                "maxRetries": 50
              },
              {
                "maxConnections": 100000,
                "maxPendingRequests": 100000,
                "maxRequests": 60000000,
                "maxRetries": 50
              }
            ]
          },
          "http2ProtocolOptions": {}
        },
        {
          "name": "service_stats",
          "type": "LOGICAL_DNS",
          "connectTimeout": "0.250s",
          "hosts": [
            {
              "socketAddress": {
                "address": "127.0.0.1",
                "portValue": 9001
              }
            }
          ]
        }
      ],
      "dynamicActiveClusters": [
        {
          "versionInfo": "0",
          "cluster": {
            "name": "default/kuard/80/da39a3ee5e",
            "type": "EDS",
            "edsClusterConfig": {
              "edsConfig": {
                "apiConfigSource": {
                  "apiType": "GRPC",
                  "grpcServices": [
                    {
                      "envoyGrpc": {
                        "clusterName": "contour"
                      }
                    }
                  ]
                }
              },
              "serviceName": "default/kuard"
            },
            "connectTimeout": "0.250s",
            "commonLbConfig": {
              "healthyPanicThreshold": {}
            }
          }
        }
      ]
    }
  }
}
```

**Comment by anapsix** [comment]
listeners seem fine to me
```
[
"0.0.0.0:8002",
"0.0.0.0:9080",
"0.0.0.0:9443"
]
```
![contour-dag](https://user-images.githubusercontent.com/1256199/47848812-8b0b8180-dde0-11e8-8961-af6ea6526461.png)

**Comment by anapsix** [comment]
Since `openssl s_client does not speak `proxy_protocol`, tried disabling proxy protocol, to simplify debugging. And .. it worked, sort of..
Pretty sure I'm hitting same issue as described in #661, when running without proxy protocol enabled, and testing HTTPS on port 9443

### with `--use-proxy-protocol`
#### port Forwarding 9443->9443
SSH to K8s node with `-L9443:127.0.0.1:9443 -L9080:127.0.0.1:9080`
```
$ /usr/local/Cellar/curl/7.62.0/bin/curl --haproxy-protocol --http1.1 --tlsv1.2 \
  -o/dev/null -4 --resolve kuard.random.io:9443:127.0.0.1 \
  --cacert ../helm/cert-manager/ssl/ca.crt \
  -sv https://kuard.random.io:9443
* Added kuard.random.io:9443:127.0.0.1 to DNS cache
* Hostname kuard.random.io was found in DNS cache
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to kuard.random.io (127.0.0.1) port 9443 (#0)
> PROXY TCP4 127.0.0.1 127.0.0.1 52771 9443
* ALPN, offering http/1.1
* Server aborted the SSL handshake
* Closing connection 0
```

#### port Forwarding 443->9443
SSH to K8s node with `-L443:127.0.0.1:9443 -L80:127.0.0.1:9080`
```
$ /usr/local/Cellar/curl/7.62.0/bin/curl --haproxy-protocol --http1.1 --tlsv1.2 \
  -o/dev/null -4 --resolve kuard.random.io:443:127.0.0.1 \
  --cacert ../helm/cert-manager/ssl/ca.crt \
  -sv https://kuard.random.io:443
* Added kuard.random.io:443:127.0.0.1 to DNS cache
* Hostname kuard.random.io was found in DNS cache
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to kuard.random.io (127.0.0.1) port 443 (#0)
> PROXY TCP4 127.0.0.1 127.0.0.1 52809 443
* ALPN, offering http/1.1
* Server aborted the SSL handshake
* Closing connection 0
```

### without `--use-proxy-protocol`
#### port Forwarding 9443->9443
SSH to K8s node with `-L9443:127.0.0.1:9443 -L9080:127.0.0.1:9080`
```
$ /usr/local/Cellar/curl/7.62.0/bin/curl --http1.1 --tlsv1.2 \
  -o/dev/null -4 --resolve kuard.random.io:9443:127.0.0.1 \
  --cacert ../helm/cert-manager/ssl/ca.crt \
  -sv https://kuard.random.io:9443
* Added kuard.random.io:9443:127.0.0.1 to DNS cache
* Hostname kuard.random.io was found in DNS cache
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to kuard.random.io (127.0.0.1) port 9443 (#0)
* ALPN, offering http/1.1
* TLS 1.2 connection using TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
* Server certificate: kuard.random.io
* Server certificate: cert-manager-ma1dev
> GET / HTTP/1.1
> Host: kuard.random.io:9443
> User-Agent: Mozilla/5.0 (compatible; MSIE 9.0; Windows NT 6.1; Trident/5.0)
> Accept: */*
> Referer:
>
< HTTP/1.1 404 Not Found
< date: Thu, 01 Nov 2018 11:59:05 GMT
< server: envoy
< content-length: 0
<
* Connection #0 to host kuard.random.io left intact
```

#### port Forwarding 443->9443
SSH to K8s node with `-L443:127.0.0.1:9443 -L80:127.0.0.1:9080`
```
$ /usr/local/Cellar/curl/7.62.0/bin/curl --http1.1 --tlsv1.2 \
  -o/dev/null -4 --resolve kuard.random.io:443:127.0.0.1 \
  --cacert ../helm/cert-manager/ssl/ca.crt \
  -sv https://kuard.random.io:443
* Added kuard.random.io:443:127.0.0.1 to DNS cache
* Hostname kuard.random.io was found in DNS cache
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to kuard.random.io (127.0.0.1) port 443 (#0)
* ALPN, offering http/1.1
* TLS 1.2 connection using TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
* Server certificate: kuard.random.io
* Server certificate: cert-manager-ma1dev
> GET / HTTP/1.1
> Host: kuard.random.io
> User-Agent: Mozilla/5.0 (compatible; MSIE 9.0; Windows NT 6.1; Trident/5.0)
> Accept: */*
> Referer:
>
< HTTP/1.1 200 OK
< content-length: 1762
< content-type: text/html
< date: Thu, 01 Nov 2018 12:01:33 GMT
< x-envoy-upstream-service-time: 0
< server: envoy
<
{ [1762 bytes data]
* Connection #0 to host kuard.random.io left intact
```

**Comment by davecheney** [comment]
That is too be expected. Look at the host header your client is sending; it ends in :9443. 

You need to make sure your client sends a host header as a browser would, preferably without a training port number

**Comment by anapsix** [comment]
sure @davecheney, that makes sense, and I'm not considering it an issue.
Yet, when working around host header with port 9443 caveat with port forwarding 443->9443, curl reports `Server aborted the SSL handshake`
```
$ /usr/local/Cellar/curl/7.62.0/bin/curl --haproxy-protocol --http1.1 --tlsv1.2 -o/dev/null -4 --resolve kuard.random.io:443:127.0.0.1 --cacert ../helm/cert-manager/ssl/ca.crt -sv https://kuard.random.io
* Added kuard.random.io:443:127.0.0.1 to DNS cache
* Hostname kuard.random.io was found in DNS cache
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to kuard.random.io (127.0.0.1) port 443 (#0)
> PROXY TCP4 127.0.0.1 127.0.0.1 52946 443
* ALPN, offering http/1.1
* Server aborted the SSL handshake
* Closing connection 0
```

At the same time HTTP works as expected
```
$ /usr/local/Cellar/curl/7.62.0/bin/curl --haproxy-protocol --http1.1 -o/dev/null -4 --resolve kuard.random.io:80:127.0.0.1 -sv http://kuard.random.io
* Added kuard.random.io:80:127.0.0.1 to DNS cache
* Hostname kuard.random.io was found in DNS cache
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to kuard.random.io (127.0.0.1) port 80 (#0)
> PROXY TCP4 127.0.0.1 127.0.0.1 52952 80
> GET / HTTP/1.1
> Host: kuard.random.io
> User-Agent: Mozilla/5.0 (compatible; MSIE 9.0; Windows NT 6.1; Trident/5.0)
> Accept: */*
> Referer:
>
< HTTP/1.1 200 OK
< content-length: 1761
< content-type: text/html
< date: Thu, 01 Nov 2018 12:46:41 GMT
< x-envoy-upstream-service-time: 0
< server: envoy
<
{ [1761 bytes data]
* Connection #0 to host kuard.random.io left intact
```

**Comment by davecheney** [comment]
Can we step back and ask what you are trying to accomplish? I understand that it’s useful to test software in a stanging environment before deploying but network forwarding jiggery-pokery is very fragile when tls and port forwarding are mixed together. 

To be clear, I do want to make sure contour is working properly in your environment, but unless you’re planning on using ssh forwarding for production traffic id like to focus on making sure contour is working properly when deployed in a routable network.

**Comment by anapsix** [comment]
I'm using a LoadBalancer in L4 mode, configured out-of-band, to proxy from it's VIP to K8s worker nodes, where Contour does what it does (routing, TLS termination, etc)
```
LB VIP 80/TCP -> K8s Worker 9080/TCP (Contour DaemonSet, using nodePort 9080) 
LB VIP 443/TCP -> K8s Worker 9443/TCP (Contour DaemonSet, using nodePort 9443)
```
This all works as expected, when proxy-protocol is disabled, with or without LoadBalancer.
Then proxy-protocol is enabled, I'm having an issue described.  
All of this "network forwarding jiggery-pokery" 😄  is happening for debug purposes only, to eliminate the LB from the process of debugging.
What I see so far, is an issue with TLS negotiation, possibly SNI related, specifically when proxy protocol is enabled.

**Comment by davecheney** [comment]
Thanks for confirming. What L4 load balancer are you using? does it understand the Amzon proxy format (i'm not sure if it's a standard or an RFC)?

**Comment by anapsix** [comment]
I believe, LoadBalancer can be safely taken out if the scope, as it does the right thing as confirmed with HTTP requests. Which is why during my tests I'm using curl to talk directly to envoy (reminder: it's listening on 9080, and 9443).
Respectfully, notice the difference with proxy protocol on, and off: https://github.com/heptio/contour/issues/793#issuecomment-435030058

At the moment, my lacking understanding of Envoy ,and it's configs prevents me from seeing an issue when looking at config dump. Though, I'm sure it's there.

**Comment by anapsix** [comment]
I've compared sorted configs with proxy enabled and disabled, and diff is unsurprising, but disappointing (as I was hoping to find a culprit..)
```diff
286c286,287
<                 ]
---
>                 ],
>                 "useProxyProto": true
375c376,377
<                 }
---
>                 },
>                 "useProxyProto": true
```

**Comment by anapsix** [comment]
The "proxy protocol" mentioned in [documentation](https://github.com/heptio/contour/blob/master/docs/proxy-proto.md) is HAPROXY PROXY protocol. Afaik, HAPROXY is the originator of the protocol. 
It's visible in tests I've posted, using curl command that Envoy does use it on non-TLS port.
@davecheney please clarify ".. Envoy does not use the HA Proxy protocol, so using curl --haproxy-protocol will not work as expected"

**Comment by davecheney** [comment]
Envoy, in the way we have it configured uses PROXY V1, https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/listener/listener.proto#envoy-api-msg-listener-filterchain. Curl may be attempting PROXY V2

**Comment by anapsix** [comment]
I've posted successful examples of using `curl --haproxy-protocol`. I swear, I did no doctor them @davecheney

**Comment by anapsix** [comment]
Regardsless, I appreciate you taking your time listening and trying to help @davecheney

**Comment by davecheney** [comment]
Closing in favour of #802. 

I'm sorry I cannot resolve this issue. The best way forward I know is to switch to Envoy's newer proxy protocol handler and hope for the best.
